Explorer/Assets/DCL/SDKComponents/MediaStream/Systems/UpdateMediaPlayerSystem.cs

```diff
@@ -49,53 +49,33 @@ protected override void Update(float t)
+
+            if (component.State != VideoState.VsError)
+                component.MediaPlayer.UpdateVolume(sceneStateProvider.IsCurrent, sdkComponent.HasVolume, sdkComponent.Volume);
 
-            void OnComplete(MediaPlayer mediaPlayer) =>
-                mediaPlayer.UpdatePlayback(sdkComponent.HasPlaying, sdkComponent.Playing);
+            HandleComponentChange(ref component, sdkComponent, sdkComponent.Url, sdkComponent.HasPlaying, sdkComponent.Playing);
+            ConsumePromise(ref component, sdkComponent.HasPlaying && sdkComponent.Playing);
         }
 
         [Query]
         private void UpdateVideoStream(ref MediaPlayerComponent component, PBVideoPlayer sdkComponent)
         {
-            UpdateMediaStream(ref component, sdkComponent, sdkComponent.Src, sdkComponent.HasVolume, sdkComponent.Volume, sdkComponent.HasPlaying, sdkComponent.Playing, OnComplete);
-            return;
+            if (!frameTimeBudget.TrySpendBudget()) return;
 
-            void OnComplete(MediaPlayer mediaPlayer) =>
-                mediaPlayer.UpdatePlayback(sdkComponent.HasPlaying, sdkComponent.Playing)
-                           .UpdatePlaybackProperties(sdkComponent);
-        }
-
-        private void UpdateMediaStream(ref MediaPlayerComponent component, IDirtyMarker marker, string url, bool hasVolume, float volume,
-            bool hasPlaying, bool playing, Action<MediaPlayer> onComplete)
-        {
             if (component.State != VideoState.VsError)
-                component.MediaPlayer.UpdateVolume(sceneStateProvider.IsCurrent, hasVolume, volume);
-
-            if (marker.IsDirty && frameTimeBudget.TrySpendBudget())
-            {
-                MediaPlayer mediaPlayer = component.MediaPlayer;
+                component.MediaPlayer.UpdateVolume(sceneStateProvider.IsCurrent, sdkComponent.HasVolume, sdkComponent.Volume);
 
-                if (component.URL != url)
-                {
-                    UpdateStreamUrl(ref component, url);
+            HandleComponentChange(ref component, sdkComponent, sdkComponent.Src, sdkComponent.HasPlaying, sdkComponent.Playing, OnPlaybackUpdate);
+            ConsumePromise(ref component, sdkComponent.HasPlaying && sdkComponent.Playing, OnMediaOpened);
 
-                    if (component.State != VideoState.VsError)
-                    {
-                        component.Cts.Cancel();
-                        component.Cts = new CancellationTokenSource();
-                        mediaPlayer.OpenMediaIfReachableAsync(webRequestController, component.URL, autoPlay: false, component.Cts.Token, OnComplete).Forget();
+            return;
 
-                        void OnComplete() =>
-                            onComplete?.Invoke(mediaPlayer);
-                    }
-                }
-                else if (component.State != VideoState.VsError)
-                    mediaPlayer.UpdatePlayback(hasPlaying, playing);
+            void OnMediaOpened(MediaPlayer mediaPlayer) =>
```

Comment:
It looks like we already pass `sdkComponent` to the function with callback, thus if you add it as an argument and convert methods to static lambdas you will be able to avoid delegate allocation.